[SYSTEMDS-3385] Integration of frontend with monitoring backend

MKehayov opened the pull request:
This change has the initial frontend application written in Angular. It is integrated with the backend so that both applications are communicating with REST requests, with this, coordinators and workers can be managed from the Angular app and real-time statistics can be displayed.

j143 commented:
Hi @MKehayov thanks a lot for the feature. you can add screenshots here too!

MKehayov commented:
Sure @j143 here - 

![Screenshot from 2022-06-20 22-04-30](https://user-images.githubusercontent.com/13037174/174671256-45928ab0-9bc0-4942-8522-b19ebe06c74e.png)
![Screenshot from 2022-06-20 22-03-43](https://user-images.githubusercontent.com/13037174/174671260-66eeb938-ea5d-4b10-96bd-b09724413a9f.png)
![Screenshot from 2022-06-20 22-02-26](https://user-images.githubusercontent.com/13037174/174671263-4d28beec-e7b4-476a-ae7a-d0e2b1daa8a1.png)
![Screenshot from 2022-06-20 22-01-55](https://user-images.githubusercontent.com/13037174/174671264-9fa71630-9c86-468f-9ea9-8e8bf350a899.png)
![Screenshot from 2022-06-20 22-01-39](https://user-images.githubusercontent.com/13037174/174671267-9ad34f14-a101-4257-881a-a67117b55da1.png)

let me know if you think further changes are required for the initial implementation.

j143 commented:
thanks @MKehayov , I will tryout this feature with some experiments and let you know how it goes.

Baunsgaard commented on src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/NodeEntityModel.java:
```diff
@@ -68,13 +83,43 @@ public void setStats(final List<BaseEntityModel> stats) {
+
+		if (_requestTypeCount != null) {
+			for(var entry : _requestTypeCount) {
+				requestTypeCountStrArr.add(String.format("{" +
```
this is not nice, make the formatting string outside the for loop (i do not trust the java compiler to fix this.)

MKehayov replied:
Done, moved logic into a separate file.

Baunsgaard commented on src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/NodeEntityModel.java:
```diff
@@ -68,13 +83,43 @@ public void setStats(final List<BaseEntityModel> stats) {
 				"\"name\": \"%s\"," +
 				"\"address\": \"%s\"," +
+				"\"isOnline\": %b," +
+				"\"jitCompileTime\": %.2f," +
```
same here, make a static string or something different.

MKehayov replied:
Same, moved logic into a separate file.

Baunsgaard commented on src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/StatsEntityModel.java:
```diff
@@ -74,25 +97,33 @@ public double getMemoryUsage() {
-					"\"duration\": \"%.2f\"," +
-					"},", instruction, count, duration));
+				heavyHittersStrArr.add(String.format("{" +
+					"\"instruction\": \"%s\"," +
```
make format string outside loop

MKehayov replied:
Done, moved logic into a separate file.

Baunsgaard commented on src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/StatsEntityModel.java:
```diff
@@ -103,21 +134,22 @@ public void setHeavyHitterInstructions(final String heavyHitterInstructionsJsonS
-				sb.append(String.format("{" +
-					"\"datetime\": %s," +
+				transferredBytesStrArr.add(String.format("{" +
+					"\"datetime\": \"%s\"," +
```
format string outside for loop

MKehayov replied:
Done, moved logic into a separate file.

Baunsgaard commented on src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/repositories/DerbyRepository.java:
```diff
@@ -43,17 +43,21 @@ public class DerbyRepository implements IRepository {
 	private static final String ENTITY_SCHEMA_CREATE_STATS_STMT = "CREATE TABLE %s " +
 			"(id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY (START WITH 1, INCREMENT BY 1), " +
 			"%s INTEGER, " +
```
this is a good example of pre-computing your format string

Baunsgaard commented on src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/services/StatsService.java:
```diff
@@ -45,13 +50,20 @@ public static BaseEntityModel getWorkerStatistics(Long id, String address) {
+			log.error("Worker offline: " + dre.getMessage());
 		} catch (Exception e) {
-			throw new RuntimeException(e);
+			log.error("Error: " + e.getMessage());
```
this is super dangerous. 
Please verify that this is the actual behavior you want, 
can you limit the exception caught to something else?
And if it really is the functionality you want add a comment saying that we do not handle the exception but just log it, and that it is fine.

MKehayov replied:
Completely agree, for now, I left it as it is with additional comments, the problem was that there is a separate thread running for worker stats gathering and if the method fails, the thread stops, and no further data is gathered from any of the workers, the way to fix it is to restart the whole app. With this, the app can continue functioning without the restart.

Baunsgaard reviewed (commented):
This looks great! but we are missing tests.

I don't think we need the file:
Package-lock.json
can this not be git ignored as well?

and tsconfic.spec.json complain about the license not being valid.
In this case just remove it and set it to ignore this file in our pom.xml

best regards
Sebastian

j143 commented:
I have only verified the angular UI, (I will test the full functionality while running the federated experiments shortly.)

1. we can customize README and add license!
2. we can remove package-lock.json. Although, the standard practice is to include in a web repo, here we can keep the code concise by removing the generated code as much as possible.
3. You can add documentation (like a small diagram of the data model if possible), so that people can easily extend it.
4. Of course, as Sebastian mentioned let's add test for each component.
5. Also, let only reveal the enough info in the logs - if need be we can extend it later.

Best,
Janardhan

MKehayov commented:
Hey, @j143 thank you for the points, my fixes on them are the following:

1. I have added README files for the angular UI and backend, a license is included in the angular one.
2. I agree, that it is a standard practice to have the package-lock in the repo, it is currently there, but let me know if I should remove it.
3. Added in the README file of the backend and the API reference can be found in the angular README.
4. I added at least one test for every component, they currently sum up to 36.
5. I removed the logging of the errors in the StatsService, let me know where else you think we should remove the logs.

Best,
Mito

j143 commented:
> I agree, that it is a standard practice to have the package-lock in the repo, it is currently there, but let me know if I should remove it.

Yes, lets remove it.

j143 commented:
I am facing issues with setting up the federated experiments, could not verify this complete feature at the moment. I will do this as soon as possible.

MKehayov commented:
Hi @j143, did you manage to run and test the app, and if not, can I help with anything?

j143 commented:
> Hi @j143, did you manage to run and test the app, and if not, can I help with anything?

can you help me set it up. my email is janardhan@apache.org

j143 commented:
Hi @MKehayov - view is not responding/updating when I am trying to add either worker/coordinator.

MKehayov commented:
Hey @j143 I found a bug, which also prevents the registering of worker/coordinator in the system, and pushed a fix for it, if this does not fix the problem for you, let me know. With the latest changes, I started an implementation of an event timeline for each worker and extended the worker dashboard itself, attached are some screenshots:
![Screenshot from 2022-07-29 13-27-58](https://user-images.githubusercontent.com/13037174/181919911-d9f5474d-4957-4438-9446-28db7035cb36.png)
![Screenshot from 2022-07-30 11-00-40](https://user-images.githubusercontent.com/13037174/181919913-d8491eb9-a45f-4625-bc9e-b4235fbbf70d.png)

j143 commented:
I tried the latest change set, 😃

1. The worker, coordinator and event timeline looks good.
2. and for `npm run test` I received the following:

![image](https://user-images.githubusercontent.com/53068787/184501593-db7bf5e5-49b7-4815-a454-cacb9866c4fd.png)

Thanks for working on this feature.✨

j143 commented:
Looks Good to me. 👍 

Thanks a lot - @MKehayov. ✨🎉
